boot, snap-bootstrap: move initramfs-mounts logic to boot pkg

anonymouse64 opened the pull request:
This is the middle ground re-implementation of https://github.com/snapcore/snapd/pull/8326.

Many of the test permutations here duplicate what's already in cmd_initramfs_mounts_test.go, and we could now perhaps delete some of the tests from snap-bootstrap as it kinda duplicates now, but I think it's better to do that later.

After this is merged I can open up the followup to actually use this in the boottest pkg and write more complete failover tests for base snaps and also managers tests.

pedronis commented on boot/boot.go:
```diff
@@ -163,16 +163,21 @@ func applicable(s snap.PlaceInfo, t snap.Type, dev Device) bool {
 	// revisions retrieves the revisions of the current snap and
 	// the try snap (only the latter might not be set), and
 	// the status of the trying snap.
+	// Note that the error could be ErrorKindTrySnap, in which case curSnap may
```
ErrorKindTrySnap doesn't exist anymore, this need to be reformulated

anonymouse64 replied:
yes I missed this thanks for spotting

pedronis commented on boot/bootstate20.go:
```diff
@@ -244,6 +253,50 @@ func (ks20 *bootState20Kernel) commit() error {
+// modeenv, but no state needs to be committed when choosing to mount a
+// kernel snap.
+func (ks20 *bootState20Kernel) chooseAndCommitSnapInitramfsMount() (sn snap.PlaceInfo, err error) {
+	err = ks20.kModeenv.loadModeenv()
```
the caller sets that in, do we need the load?

anonymouse64 replied:
hmm, I would rather have it to keep consistency with the other methods on bootState20Kernel that use the modeenv, but strictly speaking we probably don't need it right now

anonymouse64 replied:
I'm going to leave it in, but if someone feels strongly about this I can remove it

pedronis commented on boot/bootstate20.go:
```diff
@@ -302,10 +355,10 @@ func (bs20 *bootState20Base) revisions() (curSnap, trySnap snap.PlaceInfo, tryin
 	}
 
-	if bs20.modeenv.BaseStatus == TryingStatus && bs20.modeenv.TryBase != "" {
+	if bs20.modeenv.BaseStatus != DefaultStatus && bs20.modeenv.TryBase != "" {
```
can you explain this change? does it affect old code?

anonymouse64 replied:
see the commit message for this change: https://github.com/snapcore/snapd/pull/8340/commits/3b5dfcb76fe3883262cefbaf22c0553d57b1b22b

specifically, during normal running status, snapd will only want the try base snap if we are in "trying", because it will already have transitioned to "trying" during the initramfs. However when we are in the initramfs and using this same code, we only want the try base snap when we are in "try", because we haven't transitioned it yet (indeed we are in the process of doing that). So, we can relax this check a bit to just check that it's not in default status, meaning it could be in try or trying and we return it and let the higher level function work out the logic by inspecting BaseStatus themselves (which they do in both cases)

pedronis commented on boot/bootstate20.go:
```diff
@@ -373,6 +426,56 @@ func (bs20 *bootState20Base) commit() error {
+// modeenv, but no state needs to be committed when choosing to mount a
+// kernel snap.
+func (bs20 *bootState20Base) chooseAndCommitSnapInitramfsMount() (sn snap.PlaceInfo, err error) {
+	err = bs20.loadModeenv()
```
same

pedronis commented on boot/bootstate20.go:
```diff
@@ -587,3 +690,65 @@ func (bsmark *bootState20MarkSuccessful) commit() error {
+
+	// check that the current snap actually exists
+	file := curSnap.Filename()
+	snapPath := filepath.Join(dirs.SnapBlobDirUnder(filepath.Join(dirs.EarlyBootUbuntuData, "system-data")), file)
```
this needs a rootdir concept?

anonymouse64 replied:
we resolved the larger issue here in the previous PR, now we can use boot pkg exported vars directly (i.e. `EarlyWritable`), then this code becomes essentially:

```suggestion
	snapPath := filepath.Join(dirs.SnapBlobDirUnder(EarlyWritable), file)
```

is that okay, or should we still pass in a rootdir somehow?

pedronis commented on boot/bootstate20.go:
```diff
@@ -587,3 +690,65 @@ func (bsmark *bootState20MarkSuccessful) commit() error {
+			if trySnap != nil {
+				// check that the TryBase exists in ubuntu-data, if it doesn't
+				// we will fall back to using the normal snap
+				trySnapPath := filepath.Join(dirs.SnapBlobDirUnder(filepath.Join(dirs.EarlyBootUbuntuData, "system-data")), trySnap.Filename())
```
same

pedronis commented on boot/initramfs.go:
```diff
@@ -0,0 +1,72 @@
+func InitramfsRunModeChooseSnapsToMount(
+	typs []snap.Type,
+	blDir string,
+	blOpts *bootloader.Options,
```
do we strictly need the bootloader.Options ?

anonymouse64 replied:
hmm, I guess I need to check if there will be a bootloader on /boot/grub and whether we will auto-find the right one in `/EFI/ubuntu` without the option set, but I'm pretty sure we do need this

pedronis replied:
sorry, I mean to be need them passed in? or we basically know what we need (like makebootable) ?

pedronis replied:
As I suppose with decisions we made we don't need blDir either?

pedronis replied:
we need modeenv because it's used both here and by the caller

anonymouse64 replied:
ah I see what you mean, yes I suppose we could specify those just in this function instead of in snap-bootstrap

pedronis commented on boot/initramfs.go:
```diff
@@ -0,0 +1,72 @@
+		case snap.TypeBase:
+			bs := &bootState20Base{}
+			fmt.Println("bs.modeenv from top level func is", modeenv)
+			bs.modeenv = modeenv
```
can't this be done above in one go?

anonymouse64 replied:
no, because go is weird about assigning to promoted members in a struct literal, see:

```
$ go build ./boot/...
boot/initramfs.go:46:5: cannot use promoted field bootState20Modeenv.modeenv in struct literal of type bootState20Base
```

pedronis commented on boot/initramfs.go:
```diff
@@ -0,0 +1,72 @@
+		switch typ {
+		case snap.TypeBase:
+			bs := &bootState20Base{}
+			fmt.Println("bs.modeenv from top level func is", modeenv)
```
left over debugging print

pedronis commented on boot/makebootable.go:
```diff
@@ -224,15 +224,12 @@ func makeBootable20(model *asserts.Model, rootdir string, bootWith *BootableSet)
 	// copy kernel/base into the ubuntu-data partition
-	ubuntuDataMnt := filepath.Join(runMnt, "ubuntu-data")
-	snapBlobDir := dirs.SnapBlobDirUnder(filepath.Join(ubuntuDataMnt, "system-data"))
+	systemData := filepath.Join(dirs.EarlyBootUbuntuData, "system-data")
```
I forgot we do this, we need to think a bit, it's not bad in itself,  but I'm not sure having them defined in dirs is good

anonymouse64 replied:
as mentioned above, we resolved this in the previous PR so I think this will be cleaner and make more sense because makebootable will use boot pkg exported vars instead

pedronis commented on boot/errors.go:
```diff
@@ -0,0 +1,38 @@
+
+// IsTrySnapError returns true if the given error is an error resulting from
+// accessing information about the try snap or the trying status.
+func IsTrySnapError(err error) bool {
```
does this needs to be exported?

anonymouse64 replied:
no, it does not

pedronis commented on boot/bootstate20.go:
```diff
@@ -244,6 +253,50 @@ func (ks20 *bootState20Kernel) commit() error {
+	}
+
+	// first do the generic choice of which snap to use
+	first, second, err := genericEarlyBootChooseSnap(ks20, TryingStatus, "kernel")
```
seems there is a bunch of places where we need to s/EarlyBoot/Initramfs/

pedronis commented on boot/bootstate20.go:
```diff
@@ -244,6 +253,50 @@ func (ks20 *bootState20Kernel) commit() error {
+			}
+			// log message that first (try snap) wasn't trusted
+			logger.Noticef("cannot use try-kernel snap: %q is not trusted in the modeenv", first.Filename())
+			return second, nil
```
will snapd in user space do the right thing in this case?

anonymouse64 replied:
hmm? I don't understand the question as snapd in user space should never call this function since it's specifically only for the initramfs, hence it's just for `snap-bootstrap initramfs-mounts` to use

pedronis replied:
I mean we are picking a different outcome without manipulating the status,  will snapd notice this?

anonymouse64 replied:
TLDR, no snapd doesn't notice this. What we need to do actually is reboot instead of falling back here. We could manipulate the status and let snapd in user-space figure it out, but we won't necessarily get that far, because we could have incompatible kernel modules between the fallback kernel we mounted and the try kernel we booted. The other thing about doing that is that snapd would also need to detect that the mounted snap and kernel snap are out of sync and do a reboot to fix the situation, which is not ideal.

pedronis commented on boot/bootstate20.go:
```diff
@@ -587,3 +690,65 @@ func (bsmark *bootState20MarkSuccessful) commit() error {
+	file := curSnap.Filename()
+	snapPath := filepath.Join(dirs.SnapBlobDirUnder(InitramfsWritableDir), file)
+	if !osutil.FileExists(snapPath) {
+		// somehow the kernel snap on ubuntu-boot doesn't exist in ubuntu-data
```
kernel can also be base here?

anonymouse64 replied:
ah yes this comment should be generalized

pedronis commented on boot/bootstate20.go:
```diff
@@ -587,3 +690,65 @@ func (bsmark *bootState20MarkSuccessful) commit() error {
+// fallback snap. It returns both so that a higher level function can do
+// additional verification of the try snap if it wants, such as the kernel snaps
+// being verified in the modeenv, or the base snaps updating the modeenv with
+// new values for base_status, etc.
```
I notice that the comment doesn't mention the expectedTryStatus, I also wonder if that should be more something like expectFallback bool and let the caller deal with what it means

anonymouse64 replied:
done, returning expectFallback does make the usages slightly simpler

pedronis commented on boot/bootstate20.go:
```diff
@@ -137,8 +145,9 @@ func (ks20 *bootState20Kernel) revisions() (curSnap, trySnap snap.PlaceInfo, try
+	// if err is ErrNoTryKernelRef, then we will just return nil as the trySnap
 	if err != nil && err != bootloader.ErrNoTryKernelRef {
-		return nil, nil, "", fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ks20.ebl.Name(), err)
+		return bootSn, nil, "", trySnapError(fmt.Sprintf("cannot identify try kernel snap with bootloader %s: %v", ks20.ebl.Name(), err))
```
wondering if newTrySnapError(fmt, a...) would be slightly nicer?

anonymouse64 replied:
yes, I can do that easily

anonymouse64 commented:
Unfortunately this change shouldn't be done until after https://github.com/snapcore/snapd/pull/8411, so the git log will get messy here I think... if folks want me to force-push with a cleaner history I am happy to do that, but I don't want to make the comments go away if folks want to keep those here.

anonymouse64 commented:
I have addressed all of the feedback locally and also resolved conflicts with https://github.com/snapcore/snapd/pull/8411, but I still want that to be merged before continuing with this. As such, marking blocked until that PR is merged.

anonymouse64 commented:
The pre-reqs here have all landed and I have addressed all of @pedronis's feedback, so this is ready again and I have unmarked it as blocked

anonymouse64 commented:
Marking this as blocked until https://github.com/snapcore/snapd/pull/8512 is merged

pedronis commented:
apart the infinite conflicts this is/can be probably unblocked now

anonymouse64 commented:
This is ready for review again.

anonymouse64 commented on boot/bootstate20.go:
```diff
@@ -385,15 +392,20 @@ func (ks20 *bootState20Kernel) loadBootenv() error {
-	bl, err := bootloader.Find("", opts)
+
+	var opts *bootloader.Options
+	if ks20.blOpts != nil {
+		opts = ks20.blOpts
+	} else {
+		opts = &bootloader.Options{}
+	}
+	// we want extracted run kernel images for uc20
+	// TODO:UC20: the name of this flag is now confusing, as it is being
+	//            slightly abused to tell the uboot bootloader to just look
+	//            in a different directory, even when we don't have an
+	//            actual extracted kernel image for that impl
+	opts.ExtractedRunKernelImage = true
```
this section is a bit odd I'm not sure what the right thing to do here though is, should we enforce that blOpts always has ExtractedRunKernelImage set to true?

pedronis replied:
afaict this should be done only in the else above?  also probably time to find the different name and use it in a follow up. Maybe UC20RunMode ?

anonymouse64 replied:
hmm this is a weird situation. We could actually drop defining this flag for all blOpts passed to the function that are not nil as you suggest, and only set ExtractedRunKernelImage for the else case when we weren't provided a blOpts to this function, but that is a bit confusing, because currently actually the uboot implementation treats NoSlashBoot as really (NoSlashBoot | ExtractedRunKernelImage), so when InitramfsRunModeSelectSnapsToMount creates the blOpts as just NoSlashBoot, it just happens to work for uboot, but if InitramfsRunModeSelectSnapsToMount specifies NoSlashBoot and ExtractedRunKernelImage then grub works from the Initramfs, but would otherwise behave differently in userspace. So I guess all this is to say I'm really confused what the flag should be renamed to because it has conflicting usages across the bootloaders it seems.

Anyways, I'm happy to just specify ExtractedRunKernelImage in the else case here as that seems to work fine, and then clean this up more in a followup but I don't know what to do about it really.

pedronis commented on boot/initramfs.go:
```diff
@@ -0,0 +1,69 @@
+	"github.com/snapcore/snapd/snap"
+)
+
+// TODO:UC20: add install/recover mode mounts code here too?
```
my feeling on this would be no, this package is already carrying a lot of responsibility as is

anonymouse64 replied:
ok, I will remove the TODO

pedronis commented on boot/initramfs.go:
```diff
@@ -0,0 +1,69 @@
+	m := make(map[snap.Type]snap.PlaceInfo)
+	for _, typ := range typs {
+		// TODO: it would be nice to use an interface here since the methods are
+		// already organized this way...
```
the interface doesn't help with the construction, no? so it would just be the method?  that can be covered with a function var no?

```
var  select fun()  (snap.PlaceInfo, error)

select := bs....

select := bs....
```

am I missing something more?

anonymouse64 replied:
well it would be nice if the interface could also help with the construction. at this point I don't even remember how I initially solved this because I did originally have this as an interface, but doing as you show with a var for a func is fine and reduces some of the duplicative code here

pedronis commented on boot/bootstate20.go:
```diff
@@ -810,3 +924,67 @@ func (bsmark *bootState20MarkSuccessful) commit() error {
+// snap is the fallback or non-trying snap and the first snap is the try snap.
+func genericInitramfsSelectSnap(bs bootState, expectedTryStatus, typeString string) (
+	firstChoice, secondChoice snap.PlaceInfo,
+	fallbackExpected bool,
```
looking at this again, do we need the flag,  it's basically  second != nil (if no error),  I'm unsure whether the flag clarifies or obscures things. for sure here it's easier to not have the flag

bboozzoo replied:
Yeah, I agree the flag is bit confusing when jumping between this place and where the function gets called in respective `selectAndCommitSnapInitramfsMount()`

anonymouse64 replied:
we have now come full circle on this hooray

:arrow_right: :arrow_right: :arrow_down: 
:arrow_up: :rainbow:  :arrow_down: 
:arrow_up: :arrow_left: :arrow_left: 

sure I can unrefactor that argument and just have the callers do `second != nil`.

pedronis replied:
not really :) , my first proposal which probably wasn't clear, was for a flag as input for  snapTryStatus == expectedTryStatus (not output) but that doesn't work because of where the loading happens which I missed when I made the proposal

pedronis replied:
> I notice that the comment doesn't mention the expectedTryStatus, I also wonder if that should 
> be more something like expectFallback bool and let the caller deal with what it means

I wanted `expectedTryStatus string` to be `expectFallback bool`

anonymouse64 replied:
well in any case I don't have the argument here anymore so hopefully it is to everyone's tastes :-)

bboozzoo commented on boot/bootstate20.go:
```diff
@@ -525,6 +538,52 @@ func (ks20 *bootState20Kernel) commit() error {
+	}
+
+	// always try the first and fallback to the second if we fail
+	if validKernels[first.Filename()] {
```
could we use strutil.ListContain instead? This is going to be a couple of snaps at most

anonymouse64 replied:
sure

anonymouse64 commented on bootloader/bootloadertest/bootloadertest.go:
```diff
@@ -45,10 +45,19 @@ type MockBootloader struct {
 }
 
-// ensure MockBootloader(s) implement the Bootloader interfaceces
+type Mock20BootloaderSetup interface {
```
no, it can't be boottest.MockUC20RunBootenv because that already exists and is a function. What's wrong with Mock20BootloaderSetup ? the interface is not specific to run mode as far as I can tell?

anonymouse64 replied:
I will move the interface definition to boottest as you prefer however

pedronis replied:
actually I see is just used in one place atm, it might be easier to define a local interface there for now

anonymouse64 replied:
sure I can do that too

anonymouse64 commented on boot/bootstate20.go:
```diff
@@ -810,3 +924,67 @@ func (bsmark *bootState20MarkSuccessful) commit() error {
+		return nil, nil, false, fmt.Errorf("%s snap %q does not exist on ubuntu-data", typeString, file)
+	}
+
+	if err != nil && isTrySnapError(err) {
```
sure this is easier to understand and logically equivalent, applied. Thanks!

anonymouse64 commented on boot/bootstate20.go:
```diff
@@ -659,6 +718,62 @@ func (bs20 *bootState20Base) commit() error {
+		// we tried to boot a try base snap and failed, so we need to reset
+		// BaseStatus
+		bs20.modeenv.BaseStatus = DefaultStatus
+		modeenvChanged = true
```
it's already cleared in the bootState20MarkSuccessful.commit method which will be eventually called during snapd startup from boot.MarkBootSuccessful, so we don't need to clear it here. We need to clear/reset BaseStatus here though because that is used to determine other logic. 

I think as a design point, we should have the initramfs mutate as little as possible, so it feels better to not clean this here and let snapd userspace cleanup. Additionally, this logic matches the uc16 case, where the bootloader does not reset snap_try_core, i.e. see https://github.com/snapcore/pc-amd64-gadget/blob/16/grub.cfg.

pedronis commented on boot/bootstate20.go:
```diff
@@ -810,3 +921,69 @@ func (bsmark *bootState20MarkSuccessful) commit() error {
+	}
+
+	if err != nil && isTrySnapError(err) {
+		// just log that we had issues with the try snap and continue with
```
we don't seem to exercise this nor the places that use newTrySnapError

bboozzoo replied:
The check in `isTrySnapErr()` is wrong. I've pushed a patch fixing the check and extending coverage.

anonymouse64 replied:
thanks @bboozzoo

pedronis commented on boot/initramfs_test.go:
```diff
@@ -0,0 +1,536 @@
+
+	// do both the normal uc20 bootloader and the env ref bootloader
+	bloaderTable := []struct {
+		bl interface {
```
thanks

anonymouse64 commented on boot/initramfs.go:
```diff
@@ -0,0 +1,64 @@
+		switch typ {
+		case snap.TypeBase:
+			bs := &bootState20Base{}
+			bs.modeenv = modeenv
```
ah thanks!
